flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,230 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.changelog.fs;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.state.KeyGroupRange;
+import org.apache.flink.runtime.state.StreamStateHandle;
+import org.apache.flink.runtime.state.changelog.SequenceNumber;
+import org.apache.flink.runtime.state.changelog.StateChange;
+import org.apache.flink.runtime.state.changelog.StateChangelogHandleStreamImpl;
+import org.apache.flink.runtime.state.changelog.StateChangelogWriter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.changelog.fs.StateChangeSet.Status.PENDING;
+import static org.apache.flink.runtime.concurrent.FutureUtils.combineAll;
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+@NotThreadSafe
+class FsStateChangelogWriter implements StateChangelogWriter<StateChangelogHandleStreamImpl> {
+    private static final Logger LOG = LoggerFactory.getLogger(FsStateChangelogWriter.class);
+
+    private final UUID logId;
+    private final KeyGroupRange keyGroupRange;
+    private final StateChangeStore store;
+    private final NavigableMap<SequenceNumber, StateChangeSet> changeSets = new TreeMap<>();
+    private List<StateChange> activeChangeSet = new ArrayList<>();
+    private SequenceNumber lastAppendedSequenceNumber = SequenceNumber.of(0L);
+    private boolean closed;
+
+    FsStateChangelogWriter(UUID logId, KeyGroupRange keyGroupRange, StateChangeStore store) {
+        this.logId = logId;
+        this.keyGroupRange = keyGroupRange;
+        this.store = store;
+    }
+
+    @Override
+    public void append(int keyGroup, byte[] value) {
+        LOG.trace("append to {}: keyGroup={} {} bytes", logId, keyGroup, value.length);
+        checkState(!closed, "%s is closed", logId);
+        activeChangeSet.add(new StateChange(keyGroup, value));
+        // size threshold could be added to call persist when reached. considerations:
+        // 0. can actually degrade performance by amplifying number of requests
+        // 1. which range to persist?
+        // 2. how to deal with retries/aborts?
+    }
+
+    @Override
+    public SequenceNumber lastAppendedSequenceNumber() {
+        LOG.trace("query {} sqn: {}", logId, lastAppendedSequenceNumber);
+        rollover();
+        return lastAppendedSequenceNumber;
+    }
```

Comment:
Why do we need to rollover for a getter?

Reply:
Because we need to distinguish the changes submitted (via `append`) before and after this call.
The client (state backend) will later request to persiste changes addressing them by what is returned from here.

`InMemoryStateChangelogWriter` doesn't rollover here, but it increments the counter on every `append`.